staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -105,17 +106,17 @@ func (wm watchersMap) addWatcher(w *cacheWatcher, number int) {
 	wm[number] = w
 }
 
-func (wm watchersMap) deleteWatcher(number int, done func(*cacheWatcher)) {
+func (wm watchersMap) deleteWatcher(number int, drainWatcher bool, done func(bool, *cacheWatcher)) {
```

Comment:
OK - I think I finally figured out why I didn't like this.
Basically, the watchersMap doesn't use the information whether the watch should be drained or not (other than passing it to the watch itself).

So they way I would like this work is to:
- revert the changes to all watchersMap (those should just be taking func(*cacheWatcher) as before
- add a field to cacheWatcher like `shouldDrain`
- switch the `forget()` function to basically be sth like:
```
func forget(bool) {
  watcher.setShouldDrain(bool)
  deleteWatcher(...) // as called now without drain parameter
}
```

- and then the cacheWatcher::stopLocked() will be looking at the `shouldDrain` param and choosing what to do based on that